uamqp/aio/_cbs_async.py

```diff
@@ -0,0 +1,227 @@
+        self._mgmt_link = self._session.create_request_response_link_pair(
+            endpoint='$cbs',
+            on_amqp_management_open_complete=self._on_amqp_management_open_complete,
+            on_amqp_management_error=self._on_amqp_management_error,
+            status_code_field=b'status-code',
+            status_description_field=b'status-description'
+        )  # type: ManagementLink
+
+        if not auth.get_token or not is_coroutine(auth.get_token):
+            raise ValueError("get_token must be a coroutine object.")
+
+        self._auth = auth
+        self._encoding = 'UTF-8'
+        self._auth_timeout = kwargs.pop('auth_timeout', DEFAULT_AUTH_TIMEOUT)
+        self._token_put_time = None
+        self._expires_in = None
+        self._expires_on = None
+        self._token = None
+
+        self._refresh_window = None
+
+        self._token_status_code = None
+        self._token_status_description = None
+
+        self.state = CbsState.CLOSED
+        self.auth_state = CbsAuthState.Idle
+
+    async def _put_token(self, token, token_type, audience, expires_on=None):
+        # type: (str, str, str, datetime) -> None
+        message = Message(
+            value=token,
+            properties=Properties(message_id=self._mgmt_link.next_message_id),
+            application_properties={
+                CBS_NAME: audience,
+                CBS_OPERATION: CBS_PUT_TOKEN,
+                CBS_TYPE: token_type,
+                CBS_EXPIRATION: expires_on
+            }
+        )
+        await self._mgmt_link.execute_operation(
+            message,
+            self._on_execute_operation_complete,
+            timeout=self._auth_timeout,
+            operation=CBS_PUT_TOKEN,
+            type=token_type
+        )
+        self._mgmt_link.next_message_id += 1
+
+    async def _on_amqp_management_open_complete(self, management_open_result):
+        if self.state in (CbsState.CLOSED, CbsState.ERROR):
+            _LOGGER.info("Unexpected AMQP management open complete.")
+        elif self.state == CbsState.OPEN:
+            self.state = CbsState.ERROR
+            _LOGGER.info("Unexpected AMQP management open complete in OPEN,"
+                         "CBS error occurred on connection %r.", self._connection._container_id)
+        elif self.state == CbsState.OPENING:
+            self.state = CbsState.OPEN if management_open_result == ManagementOpenResult.OK else CbsState.CLOSED
+            _LOGGER.info("CBS for connection %r completed opening with status: %r",
+                         self._connection._container_id, management_open_result)
+
+    async def _on_amqp_management_error(self):
+        if self.state == CbsState.CLOSED:
+            _LOGGER.info("Unexpected AMQP error in CLOSED state.")
+        elif self.state == CbsState.OPENING:
+            self.state = CbsState.ERROR
+            await self._mgmt_link.close()
+            _LOGGER.info("CBS for connection %r completed opening with status: %r",
```

Comment:
Shouldn't this message be "failed to open"?

Reply:
updated